goose.go

```diff
@@ -1653,51 +1657,82 @@ func (ctx Ctx) goStmt(e *ast.GoStmt) coq.Expr {
+	switch usage {
+		case ExprValReturned:
+			s, ok := s.(*ast.ReturnStmt)
+			if ok {
+				return coq.NewAnon(ctx.returnExpr(s.Results)), true
+			}
+		case ExprValLoop:
+			s, ok := s.(*ast.BranchStmt)
+			if ok {
+				return coq.NewAnon(ctx.branchStmt(s)), true
+			}
+	}
+	// Some statements can handle the usage themselves, and they make sure to finalize it, too
+	switch s := s.(type) {
+	case *ast.IfStmt:
+		return ctx.ifStmt(s, []ast.Stmt{}, usage), true
+	case *ast.BlockStmt:
+		return coq.NewAnon(ctx.blockStmt(s, usage)), true
+	}
+	// For everything else, we generate the statement and possibly tell the caller
+	// that this is not yet finalized.
+	var binding coq.Binding = coq.Binding{}
 	switch s := s.(type) {
 	case *ast.ReturnStmt:
-		if c.HasNext() {
-			ctx.unsupported(c.Next(), "statement following return")
-			return coq.Binding{}
-		}
-		if loopVar != nil {
-			ctx.futureWork(s, "return in loop (use break)")
-			return coq.Binding{}
-		}
-		return coq.NewAnon(ctx.returnExpr(s.Results))
+		ctx.futureWork(s, "return in unsupported position")
 	case *ast.BranchStmt:
-		if loopVar == nil {
-			ctx.unsupported(s, "branching outside of a loop")
-		}
-		return coq.NewAnon(ctx.branchStmt(s))
+		ctx.futureWork(s, "break/continue in unsupported position")
 	case *ast.GoStmt:
-		return coq.NewAnon(ctx.goStmt(s))
+		binding = coq.NewAnon(ctx.goStmt(s))
 	case *ast.ExprStmt:
-		return coq.NewAnon(ctx.expr(s.X))
+		binding = coq.NewAnon(ctx.expr(s.X))
 	case *ast.AssignStmt:
-		return ctx.assignStmt(s, c, loopVar)
+		binding = ctx.assignStmt(s)
 	case *ast.DeclStmt:
-		return ctx.varDeclStmt(s)
+		binding = ctx.varDeclStmt(s)
 	case *ast.IncDecStmt:
-		return ctx.incDecStmt(s, loopVar)
-	case *ast.BlockStmt:
-		return coq.NewAnon(ctx.blockStmt(s, loopVar))
-	case *ast.IfStmt:
-		return ctx.ifStmt(s, c, loopVar)
+		binding = ctx.incDecStmt(s)
 	case *ast.ForStmt:
 		// note that this might be a nested loop,
 		// in which case the loop var gets replaced by the inner loop.
-		return coq.NewAnon(ctx.forStmt(s))
+		binding = coq.NewAnon(ctx.forStmt(s))
 	case *ast.RangeStmt:
-		return coq.NewAnon(ctx.rangeStmt(s))
+		binding = coq.NewAnon(ctx.rangeStmt(s))
 	case *ast.SwitchStmt:
 		ctx.todo(s, "check for switch statement")
 	case *ast.TypeSwitchStmt:
 		ctx.todo(s, "check for type switch statement")
 	default:
 		ctx.unsupported(s, "statement")
 	}
-	return coq.Binding{}
+	switch usage {
+		case ExprValLocal:
+			// Nothing to finalize.
+			return binding, true
+		default:
+			// Finalization required
+			return binding, false
+	}
+}
+
+// Translate a single statement and finalize it if needed
+func (ctx Ctx) stmt(s ast.Stmt, usage ExprValUsage) coq.Binding {
+	// stmts takes care of finalization...
+	block := ctx.stmts([]ast.Stmt{s}, usage)
+	// ...but let's prettify the result while converting to a Binding.
+	if len(block.Bindings) == 1 {
+		return block.Bindings[0]
+	} else {
+		return coq.NewAnon(block)
+	}
```

Comment:
This is actually crucial, some binders get lost when I always return `coq.NewAnon(block)` (so the comments should be updated, this is not just about prettification). But I am not sure why.